File: src/edu/uf/main/run/RunSingleThread.cpp
/*
 * RunSingleThread.cpp
 *
 *  Created on: Jul 22, 2024
 *      Author: henriquedeassis
 */

#include "RunSingleThread.h"
#include <iostream>
#include <chrono>
#include <vector>
#include <algorithm>
#include <random>
#include <string>
#include "../../control/Exec.h"
#include "../../interactable/afumigatus/Afumigatus.h"


namespace edu {
namespace uf {
namespace main {
namespace run {

using namespace chrono;
using namespace control;
using namespace afumigatus;

long RunSingleThread::eNextTime = 0;
long RunSingleThread::diffusionTime = 0;
long RunSingleThread::recruitTime = 0;

void RunSingleThread::run(
    int iterations,
    int xbin,
    int ybin,
    int zbin,
	Recruiter** recruiters,
    int recruiterCount,
    bool printLattice,
    const string& outputFile,
    int nthreads,
    PrintStat* printStat
) {
    int L[4] = {0, 1, 2, 3};
    random_device rd;
    mt19937 g(rd());

    for (int k = 0; k < iterations; k++) {
        if (k != 0) {
            shuffle(&L[0], &L[3], g);
        }
        if (Afumigatus::getTotalCells0() > 3e5) {
            return;
        }
        for (int ii : L) {
            if (ii == 0) {
            	//auto start  = high_resolution_clock::now();
                Exec::next(xbin, ybin, zbin);
                //auto end  = high_resolution_clock::now();
                //eNextTime += duration_cast<microseconds>(end - start).count();
            } else if (ii == 1) {
            	//auto start  = high_resolution_clock::now();
            	Exec::recruit(recruiters, recruiterCount);
            	//auto end  = high_resolution_clock::now();
            	//recruitTime += duration_cast<microseconds>(end - start).count();
            } else if (ii == 2) {
            	//auto start  = high_resolution_clock::now();
            	Exec::diffusion();
            	//auto end  = high_resolution_clock::now();
            	//diffusionTime += duration_cast<microseconds>(end - start).count();
            }
        }

        printStat->printStatistics(k, outputFile);
        Exec::resetCount();
    }
}

} // namespace run
} // namespace main
} // namespace uf
} // namespace edu

File: java/src/edu/uf/main/run/RunSingleThread.java
package edu.uf.main.run;

import java.io.File;
import java.util.Collections;

import edu.uf.compartments.Recruiter;
import edu.uf.control.Exec;
import edu.uf.control.MultiThreadExec;
import edu.uf.interactable.Afumigatus.Afumigatus;
import edu.uf.main.print.PrintStat;

public class RunSingleThread implements Run{
	
	public static long eNextTime = 0;
	public static long recruitTime = 0;
	public static long diffusionTime = 0;

	@Override
	public void run(
			int iterations, 
			int xbin, 
			int ybin, 
			int zbin, 
			Recruiter[] recruiters, 
			boolean printLattice, 
			File outputFile, 
			int nthreads,
			PrintStat printStat
	) throws InterruptedException {
    	
        for (int k = 0; k < iterations; k++) {
            if (k != 0)
                Collections.shuffle(L);
            if (Afumigatus.getTotalCells0() > 3e5)//1.1e7)
                return;
            for (int ii : L) {

                if (ii == 0) {
                	long start = System.currentTimeMillis();
                    Exec.next(xbin, ybin, zbin);
                    long end = System.currentTimeMillis();
                    eNextTime += (end - start);
                }else if (ii == 1) {
                	long start = System.currentTimeMillis();
                	Exec.recruit(recruiters);
                	long end = System.currentTimeMillis();
                	recruitTime += (end - start);
                }else if (ii == 2) {
                	long start = System.currentTimeMillis();
                	Exec.diffusion();
                	long end = System.currentTimeMillis();
                	diffusionTime += (end - start);
                }
            }
            
            /*if(k==360) {
            	for(int x = 0; x < xbin; x++) 
                	for(int y = 0; y < ybin; y++)
                		for(int z = 0; z < zbin; z++) {
                			//Hemopexin.getMolecule().values[0][x][y][z] = 1e5*Constants.DEFAULT_HPX_CONCENTRATION;
                			Heme.getMolecule().values[0][x][y][z] = 1*Constants.VOXEL_VOL;
                		}
            }*/

            printStat.printStatistics(k, outputFile);
            MultiThreadExec.resetCount();
        }
    }

}
